### PR TITLE
irc service: Append GitHub username and repository to realname

### PR DESCRIPTION
**Pull request by Zarthus**
This gives network and channel staff a pointer in case a github bot is at the wrong channel.

Due to how IRC works, if a specific configuration is causing issues, network staff will have no way to know who is hosting the bot (it disconnects too fast to know which channel it tried to join). Adding the github-username/repository string in the realname will allow them to investigate the situation and attempt to contact the user.

This change is minimalistic, and I have ensured that it falls back to the original value if the data is unavailable. The tests appear to be functional and I have confirmed with dummy data that this change does work.

Fixes https://github.com/github/github-services/issues/1016 if implemented.

I hope you consider this pull request.

### CODE REVIEW

**Comment by jasonrudolph** [comment]
Thanks for this pull request, @Zarthus.

In #1016, you expressed a concern about the fact that this approach will reveal the names of private repositories:

> With that in mind, there is an issue at hand. Private repositories will be visible in the realname. I think this issue is pretty trivial and it is the users own choice to enable the hook.

Thanks for pointing this out. From my perspective, if the repository owner configures their repository to publish repository events to an IRC channel, then the repository owner wants people in that channel to know that the repository exists. The repository owner should have no expectation that the repository name will remain secret. @kdaigle: Does that sound reasonable to you?

**Comment by Zarthus** [comment]
Yep, the thing to note is that the private repository is also broadcasted to the channel it is in. The only real case this is visible is for Network Staff (which, considering you're running your bot on their network, probably put a reasonable amount of trust in), or when someone knows the name of the bot and uses `/whowas botname` to retrieve the realname. 

I cannot think of any scenario where someone who doesn't know about the bot could retrieve the realname, but yes, in a way, leakage of this information is possible.

**Comment by kdaigle** [comment]
> @kdaigle: Does that sound reasonable to you?

This seems reasonable to me. I'd like @ptoomey3 or @gregose to give a 👍🏻 on that as well as the code though.

File: lib/services/irc.rb [hunk 1]
```diff
@@ -58,10 +58,11 @@ def send_messages(messages)
     rooms   = rooms.gsub(",", " ").split(" ").map{|room| room[0].chr == '#' ? room : "##{room}"}
     botname = data['nick'].to_s.empty? ? "GitHub#{rand(200)}" : data['nick'][0..16]
     command = config_boolean_true?('notice') ? 'NOTICE' : 'PRIVMSG'
+    geckos  = (defined?(sender) && defined?(repo)) ? "GitHub IRCBot - #{sender.login}/#{repo.name}" : "Github IRCBot"
```
Comment (jasonrudolph) [inline lib/services/irc.rb]: @Zarthus: Similar to the approach you proposed in https://github.com/github/github-services/issues/1016#issue-55607375, I recommend extracting this logic into an intention-revealing method.

What's a good name for the method? Perhaps `irc_realname`?

**Comment by jasonrudolph** [comment]
> The tests appear to be functional and I have confirmed with dummy data that this change does work.

@Zarthus: Even though the change is fairly minor, we'll be better off if we have an automated test to ensure that it keeps working over time. Can I talk you into adding a test in [`test/irc_test.rb`](https://github.com/github/github-services/blob/2e886f407696261bd5adfc99b16d36d5e7b50241/test/irc_test.rb)?

**Comment by Zarthus** [comment]
@jasonrudolph What exactly would the test do?  It falls back to the previous value that could never fail if the variables do not exist.

**Comment by ptoomey3** [comment]
it seems reasonable to disclose the repo if configuring IRC integration. And, the change itself seems fine. I had to lookup the IRC RFC, but it looks like anything after the colon in the `USER` command can be freeform text.

**Comment by ptoomey3** [comment]
Just to make sure I'm understanding things, how is this new disclosure that much different than what is done in something like this: https://github.com/Zarthus/github-services/blob/patch/irc-realname-alter/lib/services/irc.rb#L276.

**Comment by Zarthus** [comment]
IRC consists of rooms, which all have a different userbase.

The GitHub IRC service connects, sends a message to the room(s) it is configured to join, and parts.

IRC Network Staff cannot view this information back in their logs, and in my case, we had an user who told their GitHub service to identify to an account it does not belong to. We had no means of contacting them (we could not find out which room it joined, or which repository/user it belonged to), and it is not a good idea to ban the entire GitHub address range.

My change adds the repository and user to the realname, allowing IRC Network Staff to retrieve this information to contact the user, or file a ticket with GitHub support asking if they can disable the service ran by that user for that repo.

Here is an example of what the user sees when the GitHub IRC service is triggered:
```
(00:25:31) * zarthus-git (~zarthus-g@192.30.252.34) has joined #zarthus
(00:25:31) <zarthus-git> [irc-voice-bot] Zarthus pushed 1 new commit to master: http://git.io/vOd0N
(00:25:31) <zarthus-git> irc-voice-bot/master df2b0d4 Jos Ahrens: Fix name in CTCP VERSION
(00:25:31) * zarthus-git (~zarthus-g@192.30.252.34) has left #zarthus
```

And this is the only information IRC Staff currently can retrieve:
```
(00:46:47) zarthus-git was ~zarthus-g@192.30.252.34 * GitHub IRCBot
(00:46:47) zarthus-git actually using host 192.30.252.34
```

With my change, it would display the following:
```
(00:46:47) zarthus-git was ~zarthus-g@192.30.252.34 * GitHub IRCBot - #{sender.login}/#{repo.name} 
(00:46:47) zarthus-git actually using host 192.30.252.34
```

Giving staff a means of contact.

File: lib/services/irc.rb (hunk 1 above)
Comment (Zarthus) [inline lib/services/irc.rb]: Yes, I agree.

I've added a new commit doing exactly that :)

**Comment by Zarthus** [comment]
I think this should satisfy the feedback I've received regarding this Pull Request.  Cheers on the example, Jason :).

Let me know if there's anything else you'd like to see.

File: test/irc_test.rb [hunk 2]
```diff
@@ -291,6 +291,14 @@ def test_no_colors
     refute_match /\003/, privmsg
   end
 
+  def test_include_repo_name_and_owner_in_irc_realname
+    svc = service(:pull_request, {'room' => 'r', 'nick' => 'n'}, pull_payload)
+
+    svc.receive_pull_request
+    msgs = svc.writable_irc.string.split("\n")
+    assert_includes msgs, "USER n 8 * :GitHub IRCBot - defunkt/grit"
```
Comment (jasonrudolph) [inline test/irc_test.rb]: In the [payload used in the tests](https://github.com/github/github-services/blob/8ac127e018e2b8eedc14e3d823605cc2fbcfff1a/lib/service/events/helpers/helpers_with_meta.rb#L12-L16), the repository is "mojombo/grit". "defunkt" is the sender, not the repository owner. I think you want your assertion to look like [this](https://github.com/github/github-services/compare/2438ee27c35af9c16f3de31a1c702fa3caa5cf3f...8ac127e018e2b8eedc14e3d823605cc2fbcfff1a#diff-c5ef582e0e679b233de71e8fa43f5254R299).

```ruby
assert_includes msgs, "USER n 8 * :GitHub IRCBot - mojombo/grit"
```

And to make that work, I think you want the implementation to look more like [this](https://github.com/github/github-services/compare/2438ee27c35af9c16f3de31a1c702fa3caa5cf3f...8ac127e018e2b8eedc14e3d823605cc2fbcfff1a#diff-df7cb3c7def6051de45c5de7d5b4b8a8R144):

```ruby
def irc_realname
  repo_owner = payload["repository"]["owner"]["name"]
  repo_name = payload["repository"]["name"]

  "GitHub IRCBot - #{repo_owner}/#{repo_name}"
end
```

What do you think?

Feel free to cherry-pick any of [these commits](https://github.com/github/github-services/compare/2438ee27c35af9c16f3de31a1c702fa3caa5cf3f...8ac127e018e2b8eedc14e3d823605cc2fbcfff1a#diff-c5ef582e0e679b233de71e8fa43f5254R299) if you like.

**Comment by Zarthus** [comment]
Hey Jason, thanks for the feedback. 

I'll admit I was a bit confused at first but the advice you've given has helped clear things up!

**Comment by jasonrudolph** [comment]
> Does this look okay to you now?

The code looks right to me. :+1: 

Once @ptoomey3 weighs in with a :+1: or :-1: on the information disclosure discussion above, we'll know whether this can move forward. I anticipate a follow-up reply from him by early next week, but perhaps sooner.

**Comment by ptoomey3** [comment]
@Zarthus, thanks for clarifying above. Given that, I think it would be preferable to not disclose private repo information outside of the intended channel. I chatted with @jasonrudolph and @kdaigle about some alternative options. The consensus is that it would be nice to just list the repository ID. That doesn't disclose any sensitive information and would allow network staff to contact GitHub support if problems arise with a repositories IRC configuration. 

So, instead of setting the real name to `"GitHub IRCBot - #{repo_owner}/#{repo_name}"` we were thinking of something more like `"GitHub IRCBot (https://api.github.com/repositories/#{repo_id})"`. For a public repository, the network staff could get the result of that API endpoint and directly contact the repo owner/open an issue. For private repositories, that API endpoint will be empty for unauthorized users. But, the network staff could open a support ticket with GitHub and we could assist in resolving the issue with the problematic repository IRC configuration.

**Comment by Zarthus** [comment]
@ptoomey3,

I appreciate the concerns and alternate suggestions for the private repositories. 

It is an interesting idea, but I fear the length limit of a realname (which seems to get silently truncated to 50 characters max on the networks I've tested) may get the important data truncated, and I am not entirely sure if just the Repo ID is information and may lead to some confusion. Perhaps it would be possible to shorten the URL with git.io?

I'm rather confident that having the information I presented in a realname is not directly harmful to any repository. And seeing as it's just a name, they will never see the code - just the name of the repository. Giving away it exists, but in the end the user cannot do anything with it. (although I am aware github tries its best to avoid letting the user know it exists in the first place, in theory a malicious ircd or oper could spy on the conversation of the said channel - still obtaining this information). But I'm open to thinking of some alternatives! I've asked around on IRC and here are some suggestions we've came up with:

* The realname limit seems to be 50 characters, perhaps shorten the URL, or just include the ID (caveat: confusing for some users, "Why is this weird number in my bots name?")?
* Maybe make it include the channel name(s) rather than just the repo name.  (lots of caveats to this one, i.e. you can have multiple channels, channels can be secret, invite-only, key-protected, channels cannot easily ban, etc. I can elaborate on this further if anyone would like)
* user/repo (current)
* include just the repo owner's name (for a private repo)?
* include the repo owner's name - and the git repo ID for a private repo

Honestly, the issue is quite rare. And 99/100 times it would be due to user misconfiguration.  I think having the code smartly detect if it is private, and then showing less information is a fair compromise. (So user/repo for public repositories, and just the owner name for private ones).

**Comment by ptoomey3** [comment]
>  I think having the code smartly detect if it is private, and then showing less information is a fair compromise. (So user/repo for public repositories, and just the owner name for private ones).

That was definitely another option we were tossing around too. We liked that the API URL was at least consistent. But, I'm not totally opposed to using `user/repo` for public repos and `user` for private repos. The one downside with that approach is that it obviously doesn't narrow the problematic repo for organizations that have lots of repositories. But, that is an edge case of something that is already an edge case. So, I'm not super worried about it.

**Comment by Zarthus** [comment]
Hey guys, I made some changes based on how I would imagine the implementation would be done. - they're relatively significant as far as the size of the PR goes.

I'll commit the changes I have so far and will leave some comments on things to cover my thought process. It's entirely possible there are better ways to do it (specifically with regards to the HTTP query to GitHub's own API), but I wanted to give it a stab myself just to see how far I would come.

File: lib/services/irc.rb [hunk 3]
```diff
@@ -137,6 +137,31 @@ def irc_puts(command, debug_command=command)
...
   end
 
+  def irc_realname
+    repo_owner = payload["repository"]["owner"]["name"]
+    repo_name = payload["repository"]["name"]
+
+    "GitHub IRCBot - #{repo_owner}" + (repo_public? ? "/#{repo_name}" : "")
+  end
+
+  def repo_public?
```
Comment (ptoomey3) [inline lib/services/irc.rb]: Do we need this method? Looking at https://developer.github.com/v3/activity/events/types/ it seems like every event type has a repository hash with a `private` key. So, it would look something like `private = payload["repository"]["private"]`. 

@kdaigle, @jasonrudolph - Do you all think the `private` field will exist for all events that contain a `repository`?

Comment (Zarthus) [inline lib/services/irc.rb]: I wasn't certain. I couldn't find any other service doing a similar check, and the test payloads did not seem to contain them. 

However, I wrote a patch to verify and my tests did not fail. And it seems to function properly, I've added the commit to the PR.

Comment (jasonrudolph) [inline lib/services/irc.rb]: > Do you all think the `private` field will exist for all events that contain a `repository`?

Yes. You can see all the possible event types [here](https://github.com/github/github-services/blob/2e886f407696261bd5adfc99b16d36d5e7b50241/lib/service.rb#L79-L83). And you can see an example payload for each event type [here](https://developer.github.com/v3/activity/events/types/).

> I couldn't find any other service doing a similar check, and the test payloads did not seem to contain them.

It looks like the test payloads are incomplete. I recommend adding the `private` field to the test payload to support the the tests for this pull request. That will eliminate the need to make an API request to determine whether the repository is public or private.

Comment (Zarthus) [inline lib/services/irc.rb]: When I wrote https://github.com/Zarthus/github-services/commit/f695311650a2017d4ee519b07aef531cf238f808 the tests seemed to pass, further verified by https://github.com/Zarthus/github-services/blob/master/lib/service/events/status_helpers.rb#L89 and https://github.com/Zarthus/github-services/blob/master/lib/service/events/deployment_helpers.rb#L34 having private fields, I probably just looked over them. 

I'm not entirely sure why the commit is not showing up in this PR. Though. Is there a way to tell github to perform a re-check?

**Comment by jasonrudolph** [comment]
The diff looks good to me. :+1: 

@ptoomey3: How does this look from your persepective?

File: lib/services/irc.rb [hunk 4]
```diff
@@ -137,6 +137,14 @@ def irc_puts(command, debug_command=command)
     writable_irc.puts command
   end
 
+  def irc_realname
+    repo_owner = payload["repository"]["owner"]["name"]
+    repo_name = payload["repository"]["name"]
+    repo_private = payload["repository"]["private"]
+
+    "GitHub IRCBot - #{repo_owner}" + (repo_private ? "" : "/#{repo_name}")
```
Comment (ptoomey3) [inline lib/services/irc.rb]: Total nit, but my only suggestion here is to be a bit paranoid and explicitly check that `repo_private == false`. I don't think it would happen, but if `payload["repository"]["private"]` was ever `nil` for some reason we would default to disclosing the repo name. It would be best to default to not showing the repo name unless we are 100% sure that it is not private.

Comment (ptoomey3) [inline lib/services/irc.rb]: Sorry..that should have been `repo_private == true`

Comment (ptoomey3) [inline lib/services/irc.rb]: Bah..sorry..haven't had my morning coffee here. I would write this as:

``` ruby
"GitHub IRCBot - #{repo_owner}" + (repo_private == false ? "/#{repo_name}" : "" 
```

Comment (Zarthus) [inline lib/services/irc.rb]: Very nicely spotted. I've pushed the commit.

**Comment by Zarthus** [comment]
I've actually had to make a small adjustment. When I tried `repo_private == false ? 'show_name' : '')` the tests seemed to fail with a ruby error. (viewable: https://gist.github.com/Zarthus/9ff4a7ec0166201b951f)

I am not entirely sure what caused it.

File: lib/services/irc.rb [hunk 5]
```diff
@@ -137,6 +137,14 @@ def irc_puts(command, debug_command=command)
     writable_irc.puts command
   end
 
+  def irc_realname
+    repo_owner = payload["repository"]["owner"]["name"]
+    repo_name = payload["repository"]["name"]
+    repo_private = payload["repository"]["private"]
+
+    "GitHub IRCBot - #{repo_owner}" + (repo_private != true ? "/#{repo_name}" : "")
```
Comment (ptoomey3) [inline lib/services/irc.rb]: This doesn't work with `nil`:

``` ruby
nil != true ? "foo" : "bar"
=> "foo"
```

**Comment by Zarthus** [comment]
It took a bit, but I did find the missing `private` repo payload in PushHelpers. 

https://github.com/Zarthus/github-services/commit/605d76c8392913bec8d06968bb59cafc0f0f9d71#diff-5aab4f10d4ac88a6212411b6236f6caeR183

This should work fully now, I also added a test to check against nil. Which hopefully covers all bases.

**Comment by ptoomey3** [comment]
Sweet..thanks for the polish @Zarthus! :+1:

File: test/irc_test.rb [hunk 6]
```diff
@@ -291,6 +291,37 @@ def test_no_colors
...
+    svc = service({'room' => 'r', 'nick' => 'n'}, payload)
+
+    svc.receive_push
+    msgs = svc.writable_irc.string.split("\n")
+
+    assert_includes msgs, "USER n 8 * :GitHub IRCBot - mojombo/grit"
+  end
+
+  def test_private_repo_format_in_irc_realname
+    payload_copy = payload.clone
```
Comment (jasonrudolph) [inline test/irc_test.rb]: You don't need to make a clone of the payload. Each call to `#payload` returns a new object. See:

- https://github.com/github/github-services/blob/605d76c/test/helper.rb#L37-L40
- https://github.com/github/github-services/blob/2e886f4/lib/service/events/push_helpers.rb#L171-L226.

@Zarthus: Can I talk you into removing the `#clone` invocation from this test and the one below?

Comment (Zarthus) [inline test/irc_test.rb]: You're right. It seems that was absolutely not necessary.

**Comment by jasonrudolph** [comment]
Thanks, @Zarthus. This will go to production within the next few days. :star2:

**Comment by Zarthus** [comment]
Excellent! It was a great learning experience. Thanks for all the feedback provided during this Pull Request!
